### ISSUE TITLE
Simultaneous audio playback and recording issue (AUD-2678)

### ISSUE THREAD

**Issue by AntonDrubetskiy** [espressif/esp-adf#556]
Hi guys!
I have a task to read an audio signal from a microphone and simultaneously record it to an SD card and play it on headphones.
I took two examples as a basis for developing this application:
- sound recording on button press and playback on release [https://github.com/espressif/esp-adf/tree/542954aa5a532befc8036536adb2ddb18cc03db2/examples/audio_processing/pipeline_resample](url)
- http stream and simultaneous playback to headphones and recording to SD card [https://github.com/espressif/esp-adf/tree/master/examples/advanced_examples/http_play_and_save_to_file](url)
I created my project based on the "pipeline_resample" project and used elements from the "http_play_and_save_to_file" project to run two audio pipelines simultaneously and exchange data between them via a circular buffer.
The logic of my project is as follows: when ESP starts, it starts playing audio read from the microphone into the headphones and simultaneously recording to the SD card. In order to stop this and stop the device, press the REC button.
At the same time, I ran into the following problem: the recording is normal, the audio stream in the collars is noise.

Here is my project code:

`#include <string.h>
#include "freertos/FreeRTOS.h"
#include "freertos/task.h"
#include "freertos/timers.h"

#include "esp_log.h"
#include "sdkconfig.h"
#include "audio_element.h"
#include "audio_pipeline.h"
#include "audio_event_iface.h"
#include "audio_mem.h"
#include "audio_common.h"
#include "fatfs_stream.h"
#include "i2s_stream.h"
#include "wav_encoder.h"
#include "wav_decoder.h"
#include "board.h"
#include "filter_resample.h"
#include "esp_peripherals.h"
#include "periph_sdcard.h"
#include "periph_button.h"
#include "raw_stream.h"
#include "ringbuf.h"

static const char *TAG = "RESAMPLE_EXAMPLE";
static esp_periph_set_handle_t set;

#define RECORD_RATE         48000
#define RECORD_CHANNEL      2
#define RECORD_BITS         16

#define SAVE_FILE_RATE      16000
#define SAVE_FILE_CHANNEL   1
#define SAVE_FILE_BITS      16

#define PLAYBACK_RATE       48000
#define PLAYBACK_CHANNEL    2
#define PLAYBACK_BITS       16

static audio_element_handle_t create_filter(int source_rate, int source_channel, int dest_rate, int dest_channel, int mode)
{
    rsp_filter_cfg_t rsp_cfg = DEFAULT_RESAMPLE_FILTER_CONFIG();
    rsp_cfg.src_rate = source_rate;
    rsp_cfg.src_ch = source_channel;
    rsp_cfg.dest_rate = dest_rate;
    rsp_cfg.dest_ch = dest_channel;
    rsp_cfg.mode = mode;
    rsp_cfg.complexity = 0;
    return rsp_filter_init(&rsp_cfg);
}

static audio_element_handle_t create_fatfs_stream(int sample_rates, int bits, int channels, audio_stream_type_t type)
{
    fatfs_stream_cfg_t fatfs_cfg = FATFS_STREAM_CFG_DEFAULT();
    fatfs_cfg.type = type;
    audio_element_handle_t fatfs_stream = fatfs_stream_init(&fatfs_cfg);
    mem_assert(fatfs_stream);
    audio_element_info_t writer_info = {0};
    audio_element_getinfo(fatfs_stream, &writer_info);
    writer_info.bits = bits;
    writer_info.channels = channels;
    writer_info.sample_rates = sample_rates;
    audio_element_setinfo(fatfs_stream, &writer_info);
    return fatfs_stream;
}

static audio_element_handle_t create_i2s_stream(int sample_rates, int bits, int channels, audio_stream_type_t type)
{
    i2s_stream_cfg_t i2s_cfg = I2S_STREAM_CFG_DEFAULT();
    i2s_cfg.type = type;
#if defined CONFIG_ESP_LYRAT_MINI_V1_1_BOARD
    if (i2s_cfg.type == AUDIO_STREAM_READER) {
        i2s_cfg.i2s_port = 1;
        i2s_cfg.multi_out_num=1;
    }
#endif
    audio_element_handle_t i2s_stream = i2s_stream_init(&i2s_cfg);
    mem_assert(i2s_stream);
    audio_element_info_t i2s_info = {0};
    audio_element_getinfo(i2s_stream, &i2s_info);
    i2s_info.bits = bits;
    i2s_info.channels = channels;
    i2s_info.sample_rates = sample_rates;
    audio_element_setinfo(i2s_stream, &i2s_info);
    return i2s_stream;
}

static audio_element_handle_t create_wav_encoder()
{
    wav_encoder_cfg_t wav_cfg = DEFAULT_WAV_ENCODER_CONFIG();
    return wav_encoder_init(&wav_cfg);
}

void record_playback_task()
{
    audio_pipeline_handle_t pipeline_rec = NULL;
    audio_pipeline_handle_t pipeline_play = NULL;
    audio_pipeline_cfg_t pipeline_cfg = DEFAULT_AUDIO_PIPELINE_CONFIG();

    ESP_LOGI(TAG, "[1.1] Initialize recorder pipeline");
    pipeline_rec = audio_pipeline_init(&pipeline_cfg);
    pipeline_play = audio_pipeline_init(&pipeline_cfg);

    ESP_LOGI(TAG, "[1.2] Create audio elements for recorder pipeline");
    audio_element_handle_t i2s_reader_el = create_i2s_stream(RECORD_RATE, RECORD_BITS, RECORD_CHANNEL, AUDIO_STREAM_READER);
    audio_element_handle_t filter_downsample_el = create_filter(RECORD_RATE, RECORD_CHANNEL, SAVE_FILE_RATE, SAVE_FILE_CHANNEL, RESAMPLE_ENCODE_MODE);
    audio_element_handle_t wav_encoder_el = create_wav_encoder();
    audio_element_handle_t fatfs_writer_el = create_fatfs_stream(SAVE_FILE_RATE, SAVE_FILE_BITS, SAVE_FILE_CHANNEL, AUDIO_STREAM_WRITER);

    ESP_LOGI(TAG, "[1.3] Register audio elements to recorder pipeline");
    audio_pipeline_register(pipeline_rec, i2s_reader_el,        "i2s_reader");
	audio_pipeline_register(pipeline_rec, filter_downsample_el, "filter_downsample");
    audio_pipeline_register(pipeline_rec, wav_encoder_el,       "wav_encoder");
	audio_pipeline_register(pipeline_rec, fatfs_writer_el,      "file_writer");

    const char *link_tag_1[4] = {"i2s_reader", "filter_downsample", "wav_encoder", "file_writer"};
    audio_pipeline_link(pipeline_rec, &link_tag_1[0], 4);

    ESP_LOGI(TAG, "Setup file path to save recorded audio");
    audio_element_set_uri(fatfs_writer_el, "/sdcard/rec3.wav");

    ESP_LOGI(TAG, "[2.2] Create audio elements for playback pipeline");

    raw_stream_cfg_t raw_cfg = RAW_STREAM_CFG_DEFAULT();
    raw_cfg.type = AUDIO_STREAM_WRITER;
    audio_element_handle_t el_raw_read = raw_stream_init(&raw_cfg);

    audio_element_handle_t i2s_writer_el = create_i2s_stream(PLAYBACK_RATE, PLAYBACK_BITS, PLAYBACK_CHANNEL, AUDIO_STREAM_WRITER);

    ESP_LOGI(TAG, "[2.3] Register audio elements to playback pipeline");
	audio_pipeline_register(pipeline_play, el_raw_read, "raw");
	audio_pipeline_register(pipeline_play, i2s_writer_el, "i2s_writer");

    const char *link_tag[2] = {"raw", "i2s_writer"};
    audio_pipeline_link(pipeline_play, &link_tag[0], 2);

    esp_err_t retval;

    ESP_LOGI(TAG, "[2.4] Connect input ringbuffer of pipeline_save to http stream multi output");
    ringbuf_handle_t rb = audio_element_get_output_ringbuf(el_raw_read);

    ESP_LOGI(TAG,"%i", (int)rb);

    retval=audio_element_set_multi_output_ringbuf(i2s_reader_el, rb, 0);
    if(retval==ESP_ERR_INVALID_ARG)
    	ESP_LOGE(TAG,"ringbuf NOT create!");
    else
    	ESP_LOGI(TAG,"ringbuf create!");

    ESP_LOGI(TAG, "[ 3 ] Set up  event listener");
    audio_event_iface_cfg_t evt_cfg = AUDIO_EVENT_IFACE_DEFAULT_CFG();
    audio_event_iface_handle_t evt = audio_event_iface_init(&evt_cfg);

    audio_event_iface_set_listener(esp_periph_set_get_event_iface(set), evt);

    i2s_stream_set_clk(i2s_reader_el, RECORD_RATE, RECORD_BITS, RECORD_CHANNEL);
    i2s_stream_set_clk(i2s_writer_el, PLAYBACK_RATE, PLAYBACK_BITS, PLAYBACK_CHANNEL);

    ESP_LOGI(TAG, "Run pipelines");
    audio_pipeline_run(pipeline_rec);
    audio_pipeline_run(pipeline_play);

    while (1) {
        audio_event_iface_msg_t msg;
		esp_err_t ret = audio_event_iface_listen(evt, &msg, portMAX_DELAY);
		if (ret != ESP_OK) {
			ESP_LOGE(TAG, "[ * ] Event interface error : %d", ret);
			continue;
		}

		if (msg.cmd == PERIPH_BUTTON_PRESSED) {
			ESP_LOGE(TAG, "Button pressed!"); //using LOGE to make the log color different
			break;
		}
    }

    ESP_LOGI(TAG, "[ 4 ] Stop audio_pipeline");
    audio_pipeline_stop(pipeline_rec);
    audio_pipeline_wait_for_stop(pipeline_rec);
    audio_pipeline_terminate(pipeline_rec);
    audio_pipeline_stop(pipeline_play);
    audio_pipeline_wait_for_stop(pipeline_play);
    audio_pipeline_terminate(pipeline_play);

    audio_pipeline_unregister(pipeline_play, el_raw_read);
    audio_pipeline_unregister(pipeline_play, i2s_writer_el);

    audio_pipeline_unregister(pipeline_rec, i2s_reader_el);
    audio_pipeline_unregister(pipeline_rec, filter_downsample_el);
    audio_pipeline_unregister(pipeline_rec, wav_encoder_el);
    audio_pipeline_unregister(pipeline_rec, fatfs_writer_el);

    /* Terminate the pipeline before removing the listener */
    audio_pipeline_remove_listener(pipeline_rec);
    audio_pipeline_remove_listener(pipeline_play);

    /* Stop all periph before removing the listener */
    esp_periph_set_stop_all(set);
    audio_event_iface_remove_listener(esp_periph_set_get_event_iface(set), evt);

    /* Make sure audio_pipeline_remove_listener & audio_event_iface_remove_listener are called before destroying event_iface */
    audio_event_iface_destroy(evt);

    /* Release all resources */
    audio_element_deinit(el_raw_read);
    audio_element_deinit(i2s_writer_el);

    audio_element_deinit(i2s_reader_el);
    audio_element_deinit(filter_downsample_el);
    audio_element_deinit(wav_encoder_el);
    audio_element_deinit(fatfs_writer_el);
}

void app_main(void)
{
    esp_log_level_set("*", ESP_LOG_WARN);
    esp_log_level_set(TAG, ESP_LOG_INFO);

    // Initialize peripherals management
    esp_periph_config_t periph_cfg = DEFAULT_ESP_PERIPH_SET_CONFIG();
    set = esp_periph_set_init(&periph_cfg);

    // Initialize SD Card peripheral
    audio_board_sdcard_init(set, SD_MODE_1_LINE);

    // Initialize Button peripheral
    audio_board_key_init(set);

    // Setup audio codec
    audio_board_handle_t board_handle = audio_board_init();
    audio_hal_ctrl_codec(board_handle->audio_hal, AUDIO_HAL_CODEC_MODE_BOTH, AUDIO_HAL_CTRL_START);

    // Start record/playback task
    record_playback_task();
    esp_periph_set_destroy(set);
}
`

Help me figure out what the problem might be

**Comment by HengYongChao** [comment]
Hi, @AntonDrubetskiy 
I think you should make the demo step by step, suggest modify on an existing example(http_play_and_save_to_file) is a better choice.
```
[codec_chip]-->i2s_stream reader-->wav_encoder-->fatfs_stream-->[sdcard]
                         |
                         -->raw_stream-->i2s_stream_writer-->[codec chip]
```
1. Check  recorder works good. i2s - fatfs.
2. use `audio_element_set_multi_output_ringbuf` to   get raw from i2s stream reader and send to i2s writer to play.

Any question please reply here.

**Comment by jason-mao** [comment]
@AntonDrubetskiy Any update?

**Comment by AntonDrubetskiy** [comment]
> @AntonDrubetskiy Any update?

Since writing my post, I have not dealt with this issue. The fact is that initially I did it, as I wrote @HengYongChao , and nothing happened. Therefore, the answer @HengYongChao, unfortunately, did not help me in any way. Although, thanks a lot to him for that. This problem was solved in our team in a different way. If you are interested, then this task was performed using RasberryPi
